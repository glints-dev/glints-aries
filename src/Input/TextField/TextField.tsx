import * as React from 'react';

import classNames from 'classnames';

import {
  EyeIcon,
  EyeSlashedIcon,
  CloseCircleSolidIcon,
} from '../../General/Icon/components';

import {
  TextFieldContainer,
  TextFieldInput,
  TextFieldLabel,
  IconContainer,
  StartIconContainer,
} from './TextFieldStyle';

export const isFilled = (type: textFieldType, value: any) => {
  if (value === undefined || value === null) return false;
  if (type === 'number') return !isNaN(parseInt(value));
  return value !== '';
};

/** Use onChange to listen to input changes.
 * Use onBlur to listen to focus loss.
 * <br/>
 * Use status to set different styles for the Text Field based on status.
 * <br/>
 * Use small to set the size of the text field. The ...rest is passed to the
 * internal text field input. */
export const TextField: React.FunctionComponent<Props> = ({
  value,
  status,
  onKeyDown,
  disableTyping,
  small,
  type,
  label,
  disabled = false,
  className,
  removeFloatingLabel = false,
  allowClear = false,
  forwardedRef,
  onChange,
  startIcon,
  ...restProps
}) => {
  const [floating, setFloating] = React.useState<boolean>(false);
  const [inputType, setInputType] = React.useState<textFieldType>(type);
  const hasStartIcon = Boolean(startIcon);

  const handleShowPassword = React.useCallback(() => {
    setInputType(inputType === 'password' ? 'text' : 'password');
  }, [setInputType, inputType]);

  const handleClearInput = (e: React.MouseEvent<SVGSVGElement>) => {
    const event = Object.create(e);
    event.target.value = '';
    onChange(event);
  };

  const handleKeyDown = React.useCallback(
    (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (disableTyping) {
        e.preventDefault();
      }

      if (onKeyDown !== undefined) {
        return onKeyDown(e);
      }
    },
    [onKeyDown, disableTyping]
  );

  React.useEffect(() => {
    setFloating(isFilled(inputType, value));
  }, [setFloating, inputType, value]);

  const canShowClearIcon =
    allowClear && type === 'text' && value && value.length > 0;

  return (
    <TextFieldContainer className={classNames('aries-textfield', className)}>
      <TextFieldInput
        ref={forwardedRef}
        type={inputType}
        placeholder={removeFloatingLabel && label}
        status={status}
        disabled={disabled}
        onChange={onChange}
        onKeyDown={handleKeyDown}
        floating={floating}
        value={value}
        aria-label={label}
        small={small}
        disableTyping={disableTyping}
        hasStartIcon={hasStartIcon}
        {...restProps}
      />
      {startIcon && <StartIconContainer>{startIcon}</StartIconContainer>}
      {!removeFloatingLabel && (
        <TextFieldLabel
          data-testid="textfield-label"
          className="textfield-label"
          floating={floating}
          status={status}
          small={small}
          hasStartIcon={hasStartIcon}
        >
          {label}
        </TextFieldLabel>
      )}
      {canShowClearIcon && (
        <IconContainer>
          <CloseCircleSolidIcon
            data-testid="clear-button"
            onClick={handleClearInput}
          />
        </IconContainer>
      )}
      {type === 'password' && (
        <IconContainer onClick={handleShowPassword}>
          {inputType === 'password' ? (
            <EyeSlashedIcon data-testid="invisible-icon" />
          ) : (
            <EyeIcon />
          )}
        </IconContainer>
      )}
    </TextFieldContainer>
  );
};

export type textFieldType = 'text' | 'password' | 'number';

export interface Props
  extends React.ComponentPropsWithoutRef<typeof TextFieldInput> {
  type: textFieldType;
  /** Placeholder for the text field. */
  label: string;
  disabled?: boolean;
  className?: string;
  removeFloatingLabel?: boolean;
  /** If true, removes the input content with clear icon. Works for type "text" only. */
  allowClear?: boolean;
  /** Minimum value of the "number" type input */
  min?: number;
  /** Maximum value of the "number" type input. */
  max?: number;
  /** Step value for the "number" type input */
  step?: number;
  forwardedRef?: React.RefObject<HTMLInputElement>;
  /** Icon at the start of the input element. */
  startIcon?: React.ReactNode;
}

const forwardRef = (props: Props, ref: React.RefObject<HTMLInputElement>) => (
  <TextField {...props} forwardedRef={ref} />
);

forwardRef.displayName = TextField.name;

export default React.forwardRef(forwardRef);
