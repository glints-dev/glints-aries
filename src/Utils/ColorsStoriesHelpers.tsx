import React from 'react';
import styled from 'styled-components';

export const ColorSampler = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-evenly;
`;

export interface ColorBoxProps {
  bgColor: string;
  fontColor: string;
}

const StyledColorSample = styled.div`
  margin: 1em;
  flex: '1 20%';
`;

const ColorBox = styled.div<ColorBoxProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 120px;
  height: 80px;
  margin: 1em auto;
  background-color: ${({ bgColor }) => bgColor};
  font-size: 14px;
  font-weight: bold;
  color: ${({ fontColor }) => fontColor};
  box-shadow: 0 6px 10px 0 rgba(0, 0, 0, 0.15);
`;

const ColorLabel = styled.code`
  display: flex;
  justify-content: center;
  font-size: 16px;
`;

export interface ColorSampleProps extends ColorBoxProps {
  name: string;
}

export const ColorSample: React.FunctionComponent<ColorSampleProps> = ({
  bgColor,
  fontColor,
  name,
}) => (
  <StyledColorSample>
    <ColorBox bgColor={bgColor} fontColor={fontColor}>
      <code>{bgColor}</code>
    </ColorBox>
    <ColorLabel>{name}</ColorLabel>
  </StyledColorSample>
);
