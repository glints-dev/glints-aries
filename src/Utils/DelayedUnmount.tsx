import React, { useLayoutEffect, useState } from 'react';

interface Props {
  mount: boolean;
  delay?: number;
}

export const DelayedUnmount: React.FC<Props> = ({
  children,
  mount,
  delay = 300,
}) => {
  const [shouldRender, setShouldRender] = useState(false);

  useLayoutEffect(
    function setToRender() {
      if (mount) setShouldRender(true);
    },
    [mount]
  );

  useLayoutEffect(() => {
    if (!mount) {
      const timeout = setTimeout(() => setShouldRender(false), delay);
      return () => clearTimeout(timeout);
    }
  }, [mount, delay]);
  return shouldRender && <>{children}</>;
};
