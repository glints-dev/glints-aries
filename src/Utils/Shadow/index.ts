const up1 = '0 -2px 4px 0 rgba(0, 0, 0, 0.15)';
const down1 = '0 2px 4px 0 rgba(0, 0, 0, 0.15)';

const up2 = '0 -4px 8px 0 rgba(0, 0, 0, 0.15)';
const down2 = '0 4px 8px 0 rgba(0, 0, 0, 0.15)';

const up3 = '0 -6px 12px 0 rgba(0, 0, 0, 0.10)';
const down3 = '0 6px 12px 0 rgba(0, 0, 0, 0.10)';

const up4 = '0 -12px 24px 0 rgba(0, 0, 0, 0.10)';
const down4 = '0 12px 24px 0 rgba(0, 0, 0, 0.10)';

export const Shadow = {
  up1,
  down1,
  up2,
  down2,
  up3,
  down3,
  up4,
  down4,
} as const;

export type ShadowValueType = typeof Shadow[keyof typeof Shadow];

export type ShadowKeyType = keyof typeof Shadow;
