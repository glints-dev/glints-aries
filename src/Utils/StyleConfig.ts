export const Variant = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  GHOST: 'ghost',
  LINK: 'link',
};

export const TabVariant = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical',
};

export const PsychedelicTheme = {
  BLUE_DOMINANT: 'blue-dominant',
  BLUE_DEFAULT: 'blue-default',
  RED_DOMINANT: 'red-dominant',
  RED_DEFAULT: 'red-default',
  YELLOW_DOMINANT: 'yellow-dominant',
  YELLOW_DEFAULT: 'yellow-default',
};

export const Theme = {
  RED: 'red',
  YELLOW: 'yellow',
  BLUE: 'blue',
  BLUE_RED: 'blue-red',
  WHITE: 'white',
  GREY: 'grey',
};

/* Screen size library of Glints
 * Usage:
 * @media ${Device.mobileL} { ... }
 */

export const ScreenSize = {
  mobileS: 320,
  mobileM: 480,
  mobileL: 640,
  tablet: 768,
  desktopS: 1024,
  desktopM: 1260,
  desktopL: 1440,
};

export const Device = {
  mobileS: `(max-width: ${ScreenSize.mobileS}px)`,
  mobileM: `(max-width: ${ScreenSize.mobileM}px)`,
  mobileL: `(max-width: ${ScreenSize.mobileL}px)`,
  tablet: `(max-width: ${ScreenSize.tablet}px)`,
  desktopS: `(max-width: ${ScreenSize.desktopS}px)`,
  desktopM: `(max-width: ${ScreenSize.desktopM}px)`,
  desktopL: `(max-width: ${ScreenSize.desktopL}px)`,
};
