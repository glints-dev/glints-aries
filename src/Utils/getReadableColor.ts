//https://stackoverflow.com/a/35970186
export const getReadableColor = (hex: string) => {
  if (hex.indexOf('#') === 0) {
    hex = hex.slice(1);
  }
  const r = parseInt(hex.slice(0, 2), 16);
  const g = parseInt(hex.slice(2, 4), 16);
  const b = parseInt(hex.slice(4, 6), 16);
  return r * 0.299 + g * 0.587 + b * 0.114 > 186 ? '#000000' : '#FFFFFF';
};
