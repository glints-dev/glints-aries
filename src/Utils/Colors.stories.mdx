import { ColorItem, ColorPalette, Meta } from '@storybook/addon-docs';
import { PrimaryColor, Greyscale, SecondaryColor } from './Colors';
import { getReadableColor } from './getReadableColor';
import { ColorSampler, ColorSample } from './ColorsStoriesHelpers';
import * as Blue from '../@next/utilities/colors/blue';
import * as Green from '../@next/utilities/colors/green';
import * as Neutral from '../@next/utilities/colors/neutral';
import * as Orange from '../@next/utilities/colors/orange';
import * as Red from '../@next/utilities/colors/red';
import * as Yellow from '../@next/utilities/colors/yellow';

<Meta title="Utilities/Colors" />

# Color

```
import { PrimaryColor, Greyscale } from 'glints-aries';

const MyComponent = styled.div`
  background-color: ${PrimaryColor.glintsyellow};
  color: ${PrimaryColor.grey};
`;
```

## PrimaryColor

<ColorSampler>
  {Object.keys(PrimaryColor).map(colorName => (
    <ColorSample
      key={colorName}
      bgColor={PrimaryColor[colorName]}
      fontColor={getReadableColor(PrimaryColor[colorName])}
      name={colorName}
    />
  ))}
</ColorSampler>

## Greyscale

<ColorSampler>
  {Object.keys(Greyscale).map(colorName => (
    <ColorSample
      key={colorName}
      bgColor={Greyscale[colorName]}
      fontColor={getReadableColor(Greyscale[colorName])}
      name={colorName}
    />
  ))}
</ColorSampler>

## SecondaryColor

<ColorSampler>
  {Object.keys(SecondaryColor).map(colorName => (
    <ColorSample
      key={colorName}
      bgColor={SecondaryColor[colorName]}
      fontColor={getReadableColor(SecondaryColor[colorName])}
      name={colorName}
    />
  ))}
</ColorSampler>

# @next

## Brand Colors

<ColorSampler>
  <ColorSample key={Yellow.Brand} bgColor={Yellow.Brand} name="Yellow.Brand" />
  <ColorSample key={Red.Brand} bgColor={Red.Brand} name="Red.Brand" />
  <ColorSample key={Blue.Brand} bgColor={Blue.Brand} name="Blue.Brand" />
  <ColorSample key={Orange.Brand} bgColor={Orange.Brand} name="Orange.Brand" />
</ColorSampler>

## Product Colors

<ColorPalette>
  <ColorItem
    title="Blue"
    subtitle="Blue.CodeName"
    colors={Object.fromEntries(Object.entries(Blue).slice(1, 7).sort(
      (a, b) => parseInt(a[0].slice(1)) - parseInt(b[0].slice(1))
    ))}
  />
  <ColorItem
    title="Green"
    subtitle="Green.CodeName"
    colors={Object.fromEntries(Object.entries(Green).slice(0, 5).sort(
      (a, b) => parseInt(a[0].slice(1)) - parseInt(b[0].slice(1))
    ))}
  />
  <ColorItem
    title="Neutral Whites"
    subtitle="Neutral.CodeName"
    colors={Object.fromEntries(Object.entries(Neutral).sort(
      (a, b) => parseInt(a[0].slice(1)) - parseInt(b[0].slice(1))
    ).slice(5, 10))}
  />
  <ColorItem
    title="Neutral Blacks"
    subtitle="Neutral.CodeName"
    colors={Object.fromEntries(Object.entries(Neutral).sort(
      (a, b) => parseInt(a[0].slice(1)) - parseInt(b[0].slice(1))
    ).slice(0, 5))}
  />
  <ColorItem
    title="Orange"
    subtitle="Orange.CodeName"
    colors={Object.fromEntries(Object.entries(Orange).slice(1, 6).sort(
      (a, b) => parseInt(a[0].slice(1)) - parseInt(b[0].slice(1))
    ))}
  />
  <ColorItem
    title="Red"
    subtitle="Red.CodeName"
    colors={Object.fromEntries(Object.entries(Red).slice(0, 5).sort(
      (a, b) => parseInt(a[0].slice(1)) - parseInt(b[0].slice(1))
    ))}
  />
  <ColorItem
    title="Yellow"
    subtitle="Yellow.CodeName"
    colors={Object.fromEntries(Object.entries(Yellow).slice(1, 6))}
  />
</ColorPalette>
