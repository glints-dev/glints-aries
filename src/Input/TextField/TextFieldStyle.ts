import styled from 'styled-components';
import { SecondaryColor, Greyscale } from '../../Utils/Colors';

const startIconPadding = '44px';

export const IconContainer = styled.div`
  position: absolute;
  display: flex;
  right: 1em;
  cursor: pointer;

  svg {
    fill: #666666;
  }

  &:hover svg {
    fill: ${Greyscale.black};
  }
`;

export const StartIconContainer = styled.div`
  position: absolute;
  display: flex;
  left: 16px;

  svg {
    fill: #666666;
  }
`;

export const TextFieldContainer = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  flex-grow: 1;
`;

export const TextFieldLabel = styled.label<TextFieldLabelProps>`
  position: absolute;
  left: ${({ small }) => (small ? '16px' : '22px')};
  margin-right: 1px;
  overflow: hidden;
  color: #666666;
  transition: all .2s;
  pointer-events: none;
  font-weight: 400;
  font-size: ${({ small }) => (small ? '14px' : '16px')};

  ${({ hasStartIcon }) => {
    if (hasStartIcon) {
      return `
        left: ${startIconPadding};
      `;
    }
  }}
  
  ${({ floating }) => {
    if (floating) {
      return `
        padding: 0 5px;
        top: 1em;
        transform: translate3d(-15px, -20px, 0);
        transition: all .2s;
        font-size: 12px;
        background-color: ${Greyscale.white};
      `;
    }
  }}
  
  ${({ floating, small }) => {
    if (floating && small) {
      return `
        transform: translate3d(-10px, -20px, 0);
      `;
    }
  }}
`;

type statusType = 'success' | 'error';

interface TextFieldLabelProps {
  small: boolean;
  floating: boolean;
  status: statusType;
  hasStartIcon?: boolean;
}

export const TextFieldInput = styled.input<TextFieldInputProps>`
  position: relative;
  width: 100%;
  outline: none;
  font-size: ${({ small }) => (small ? '14px' : '16px')};
  padding: ${({ small }) => (small ? '8px 16px' : '12px 16px')};
  border: unset;
  border-radius: ${({ small }) => (small ? '4px' : '8px')};
  transition: all .5s;
  background-color: ${({ status }) =>
    status === 'error' ? ' rgba(236, 39, 43, 0.08)' : 'rgb(1, 126, 183, 0.08)'};

  ${({ disableTyping }) => {
    if (disableTyping) {
      return `
        color: transparent;
        text-shadow: 0 0 0 black;
        cursor: pointer;
      `;
    }
  }}

  ${({ hasStartIcon }) => {
    if (hasStartIcon) {
      return `
        padding-left: ${startIconPadding};
      `;
    }
  }}
  
  &:hover {
    transition: all .5s;
    background-color: ${({ status }) =>
      status === 'error'
        ? ' rgba(236, 39, 43, 0.08)'
        : 'rgb(1, 126, 183, 0.16)'};

    ~ ${TextFieldLabel} {
      color: #666666;
    }
  }

  &:focus {
    background-color: ${({ status }) =>
      status === 'error'
        ? ' rgba(236, 39, 43, 0.08)'
        : 'rgb(1, 126, 183, 0.24)'};

    ${({ hasStartIcon }) => {
      if (hasStartIcon) {
        return `
          + ${StartIconContainer} {
            svg {
              fill: ${SecondaryColor.actionblue};
            }
          }
        `;
      }
    }}

    ~ ${TextFieldLabel} {
      padding: 0 5px;
      top: 1em;
      transform: ${({ small }) =>
        small
          ? 'translate3d(-10px, -20px, 0)'
          : 'translate3d(-15px, -20px, 0)'};
      transition: all .2s;
      font-size: 12px;
      right: auto;
      background-color: ${Greyscale.white};
    }
  }

  &:disabled {
    cursor: not-allowed;
    background: ${SecondaryColor.whitesmoke};

    + ${TextFieldLabel} {
      background: transparent;
      color: ${SecondaryColor.grey};
    }
  }
`;

export interface TextFieldInputProps {
  small?: boolean;
  status?: statusType;
  disableTyping?: boolean;
  floating?: boolean;
  hasStartIcon?: boolean;
}
