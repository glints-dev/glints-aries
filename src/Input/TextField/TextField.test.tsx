import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TextField, { textFieldType, isFilled } from './TextField';
import { SecondaryColor } from '../../Utils/Colors';
import SearchIcon from '../../General/Icon/components/SearchIcon';

const props = {
  type: 'text' as textFieldType,
  label: 'Username',
  onChange: jest.fn().mockImplementation(event => event.target.value),
};

const inputValue = 'glintsaries';
const inputTypes: textFieldType[] = ['text', 'password', 'number'];
const emptyValueArray = [undefined, ''];
const valueArrayMap = {
  text: [inputValue, ' '],
  password: [inputValue, ' '],
  number: [0, 123],
};

function setupTextField(otherProps?: any) {
  const { getByLabelText, ...utils } = render(
    <TextField
      type={props.type}
      label={props.label}
      onChange={props.onChange}
      {...otherProps}
    />
  );
  const textFieldInput = getByLabelText(props.label) as HTMLInputElement;
  return { textFieldInput, utils };
}

function setupTextFieldWithChangeEvent(otherProps?: any) {
  const { textFieldInput } = setupTextField(otherProps);
  userEvent.type(textFieldInput, inputValue);
  return { textFieldInput };
}

it('<TextField> should render with a label of Username and type of text', () => {
  const TextFieldSnapshot = renderer
    .create(<TextField type={props.type} label={props.label} />)
    .toJSON();
  expect(TextFieldSnapshot).toMatchSnapshot();
});

describe('when it is rendered', () => {
  it('should display the correct input label', () => {
    const { textFieldInput } = setupTextField();
    expect(textFieldInput).toBeTruthy();
  });

  it('should have the correct type', () => {
    const { textFieldInput } = setupTextField();
    expect(textFieldInput.type).toEqual(props.type);
  });

  it('should have an empty string as its value', () => {
    const { textFieldInput } = setupTextField();
    expect(textFieldInput.value).toEqual('');
  });
});

describe('when a value is entered', () => {
  it('should call onChange for each input', () => {
    setupTextFieldWithChangeEvent();
    expect(props.onChange).toHaveBeenCalledTimes(inputValue.length);
  });

  it('should return the input value when onChange is passed a function: event => event.target.value', () => {
    setupTextFieldWithChangeEvent();
    expect(props.onChange.mock.results.slice(-1)[0].value).toEqual(inputValue);
  });

  it('should display the correct value', () => {
    const { textFieldInput } = setupTextFieldWithChangeEvent();
    expect(textFieldInput.value).toEqual(inputValue);
  });
});

describe('when disabled is true', () => {
  it('should display cursor: not-allowed', () => {
    const { textFieldInput } = setupTextField({ disabled: true });
    expect(textFieldInput).toHaveStyle('cursor: not-allowed;');
  });

  it('should be disabled', () => {
    const { textFieldInput } = setupTextField({ disabled: true });
    expect(textFieldInput).toBeDisabled();
  });
});

describe('when small is:', () => {
  it('true, it should display the correct size', () => {
    const { textFieldInput } = setupTextField({ small: true });
    expect(textFieldInput).toHaveStyle('padding: 8px 16px;');
  });

  it('false, it should display the correct size', () => {
    const { textFieldInput } = setupTextField();
    expect(textFieldInput).toHaveStyle('padding: 12px 16px;');
  });
});

describe('when disableTyping is true', () => {
  it('should display cursor: pointer', () => {
    const { textFieldInput } = setupTextField({ disableTyping: true });
    expect(textFieldInput).toHaveStyle('cursor: pointer;');
  });

  it('should not be able to input a value', () => {
    const { textFieldInput } = setupTextField({
      disableTyping: true,
    });
    userEvent.type(textFieldInput, inputValue);
    expect(textFieldInput.value).toBe('');
  });
});

describe('when removeFloatingLabel is true', () => {
  it('should remove the floating label', () => {
    const { textFieldInput } = setupTextField({ removeFloatingLabel: true });
    const textFieldLabel = textFieldInput.parentElement.querySelector(
      '[data-testid="textfield-label"]'
    ) as HTMLLabelElement;
    expect(textFieldInput.parentElement).not.toContainElement(textFieldLabel);
  });

  it('should display a placeholder', () => {
    const { utils } = setupTextField({ removeFloatingLabel: true });
    const textFieldInput = utils.getByPlaceholderText(props.label);
    expect(textFieldInput).toBeTruthy();
  });
});

describe('<TextField /> forwards ref to underlying input element', () => {
  const label = 'forward-ref';

  test('ref is being forwarded correctly', () => {
    const ref = React.createRef<HTMLInputElement>();
    const { queryByLabelText } = render(
      <TextField ref={ref} label={label} type="text" />
    );
    const textInput = queryByLabelText(label);
    expect(textInput).toEqual(ref.current);
  });

  test('underlying textarea can be focused/blurred through ref', () => {
    const ref = React.createRef<HTMLInputElement>();
    const { queryByLabelText } = render(
      <TextField ref={ref} label={label} type="text" />
    );
    const textInput = queryByLabelText(label);
    expect(textInput).not.toHaveFocus();

    ref.current.focus();
    expect(textInput).toHaveFocus();

    ref.current.blur();
    expect(textInput).not.toHaveFocus();
  });
});

describe('when an empty value is passed to', () => {
  test('isFilled, it should return false', () => {
    inputTypes.forEach(type => {
      emptyValueArray.forEach(emptyValue => {
        expect(isFilled(type, emptyValue)).toBe(false);
      });
    });
  });

  test('value prop, label should not be floating', () => {
    inputTypes.forEach(type => {
      emptyValueArray.forEach((emptyValue, index) => {
        const { getAllByTestId } = render(
          <TextField
            value={emptyValue}
            label="label"
            type={type}
            onChange={jest.fn()}
          />
        );
        const textFieldLabel = getAllByTestId('textfield-label')[index];
        expect(textFieldLabel).not.toHaveStyle(
          `color: ${SecondaryColor.black}`
        );
      });
    });
  });
});

describe('when a non-empty value is passed to', () => {
  test('isFilled, it should return true', () => {
    inputTypes.forEach(type => {
      const valueArray = valueArrayMap[type];
      valueArray.forEach((value: any) => {
        expect(isFilled(type, value)).toBe(true);
      });
    });
  });

  test('value prop, label should be floating', () => {
    inputTypes.forEach(type => {
      const valueArray = valueArrayMap[type];
      valueArray.forEach((value: any, index: number) => {
        const { getAllByTestId } = render(
          <TextField
            value={value}
            label="label"
            type={type}
            onChange={jest.fn()}
          />
        );
        const textFieldLabel = getAllByTestId('textfield-label')[index];
        expect(textFieldLabel).toHaveStyle(`color: #666666`);
      });
    });
  });
});

describe('when allowClear is true and type is text', () => {
  describe('and type is text', () => {
    it('should not show the clear icon when value is empty', async () => {
      const { queryByTestId } = render(
        <TextField
          label="label"
          type="text"
          allowClear={true}
          onChange={jest.fn()}
        />
      );

      const clearButton = queryByTestId('clear-button');
      expect(clearButton).not.toBeInTheDocument();
    });

    it('should show the clear icon when value is not empty', () => {
      const { queryByTestId } = render(
        <TextField
          label="label"
          type="text"
          value="123"
          allowClear={true}
          onChange={jest.fn()}
        />
      );

      const clearButton = queryByTestId('clear-button');
      expect(clearButton).toBeInTheDocument();
    });

    it('should clear the input value and trigger onChange when click clear button', () => {
      const onChange = jest.fn();

      const { queryByTestId } = render(
        <TextField
          label="label"
          type="text"
          value="123"
          allowClear={true}
          onChange={onChange}
        />
      );

      const clearButton = queryByTestId('clear-button');
      userEvent.click(clearButton);
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  describe('and type is password', () => {
    it('should show the invisible icon but not clear icon', () => {
      const { queryByTestId } = render(
        <TextField
          label="label"
          type="password"
          value="123"
          allowClear={true}
          onChange={jest.fn()}
        />
      );

      const clearButton = queryByTestId('clear-button');
      expect(clearButton).not.toBeInTheDocument();

      const invisibleIcon = queryByTestId('invisible-icon');
      expect(invisibleIcon).toBeInTheDocument();
    });
  });
});

describe('when startIcon has an icon value', () => {
  it('should match the snapshot when start icon is included inside the text box', () => {
    const { asFragment } = render(
      <TextField
        type={props.type}
        label={props.label}
        onChange={props.onChange}
        startIcon={<SearchIcon />}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
