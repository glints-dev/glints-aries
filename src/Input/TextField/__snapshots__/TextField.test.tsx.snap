// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField> should render with a label of Username and type of text 1`] = `
<div
  className="TextFieldStyle__TextFieldContainer-sc-5sd02x-2 dIeDvL aries-textfield"
>
  <input
    aria-label="Username"
    className="TextFieldStyle__TextFieldInput-sc-5sd02x-4 eOVSMW"
    onKeyDown={[Function]}
    type="text"
  />
  <label
    className="TextFieldStyle__TextFieldLabel-sc-5sd02x-3 eIevZv textfield-label"
    data-testid="textfield-label"
  >
    Username
  </label>
</div>
`;

exports[`when startIcon has an icon value should match the snapshot when start icon is included inside the text box 1`] = `
<DocumentFragment>
  <div
    class="TextFieldStyle__TextFieldContainer-sc-5sd02x-2 dIeDvL aries-textfield"
  >
    <input
      aria-label="Username"
      class="TextFieldStyle__TextFieldInput-sc-5sd02x-4 edQWCQ"
      type="text"
      value=""
    />
    <div
      class="TextFieldStyle__StartIconContainer-sc-5sd02x-1 gJuCgc"
    >
      <svg
        class="IconStyle__VerticalCenteredSvg-x7d280-0 cJSNzi"
        data-testid="icon-svg"
        fill="currentColor"
        height="1em"
        viewBox="0 0 100 100"
        width="1em"
      >
        <path
          d="M70 43c0-14.886-12.114-27-27-27S16 28.114 16 43s12.114 27 27 27 27-12.114 27-27zm30 49.308c0 4.206-3.486 7.692-7.692 7.692-2.044 0-4.027-.841-5.409-2.284L66.286 77.163c-7.031 4.868-15.445 7.452-23.978 7.452C18.93 84.615 0 65.685 0 42.308 0 18.93 18.93 0 42.308 0c23.377 0 42.307 18.93 42.307 42.308 0 8.533-2.584 16.947-7.452 23.978L97.776 86.9C99.16 88.281 100 90.264 100 92.308z"
        />
      </svg>
    </div>
    <label
      class="TextFieldStyle__TextFieldLabel-sc-5sd02x-3 gtkogB textfield-label"
      data-testid="textfield-label"
    >
      Username
    </label>
  </div>
</DocumentFragment>
`;
